File: js/packages/jobs/projection.ts
import type { Job, JobEvent, JobState } from "./types.ts";

const TERMINAL_STATES = new Set<JobState>([
  "completed",
  "failed",
  "cancelled",
  "expired",
  "dead",
  "dismissed",
]);

export function isTerminal(state: JobState): boolean {
  return TERMINAL_STATES.has(state);
}

export function jobFromWorkEvent<TPayload = unknown, TResult = unknown>(
  event: JobEvent<TPayload, TResult>,
): Job<TPayload, TResult> | undefined {
  switch (event.eventType) {
    case "created":
    case "retried":
      return seedJobFromEvent(event);
    default:
      return undefined;
  }
}

export function reduceJobEvent<TPayload = unknown, TResult = unknown>(
  current: Job<TPayload, TResult> | undefined,
  event: JobEvent<TPayload, TResult>,
): Job<TPayload, TResult> | undefined {
  if (!current) {
    if (event.eventType !== "created") {
      return undefined;
    }

    return seedJobFromEvent(event);
  }

  if (!isLegalTransition(current, event)) {
    return current;
  }

  const next: Job<TPayload, TResult> = {
    ...current,
    state: event.state,
    updatedAt: event.timestamp,
    tries: event.tries,
    ...(event.maxTries !== undefined ? { maxTries: event.maxTries } : {}),
  };

  switch (event.eventType) {
    case "started":
      next.startedAt = event.timestamp;
      break;
    case "progress":
      next.progress = event.progress;
      break;
    case "logged":
      next.logs = [...(current.logs ?? []), ...(event.logs ?? [])];
      break;
    case "completed":
      next.result = event.result;
      next.completedAt = event.timestamp;
      break;
    case "failed":
    case "cancelled":
    case "expired":
    case "dead":
    case "dismissed":
      next.lastError = event.error;
      next.completedAt = event.timestamp;
      break;
    case "retry":
      next.lastError = event.error;
      break;
    case "retried":
      if (event.payload !== undefined) {
        next.payload = event.payload;
      }
      if (event.deadline !== undefined) {
        next.deadline = event.deadline;
      }
      next.result = undefined;
      next.completedAt = undefined;
      next.startedAt = undefined;
      next.lastError = undefined;
      next.progress = undefined;
      next.logs = undefined;
      next.tries = 0;
      break;
    case "created":
      break;
  }

  return next;
}

function isLegalTransition<TPayload, TResult>(
  current: Job<TPayload, TResult>,
  event: JobEvent<TPayload, TResult>,
): boolean {
  switch (event.eventType) {
    case "created":
      return false;
    case "started":
      return (current.state === "pending" || current.state === "retry") &&
        event.previousState === current.state;
    case "retry":
    case "progress":
    case "logged":
    case "completed":
    case "failed":
      return current.state === "active" && event.previousState === "active";
    case "cancelled":
    case "expired":
      return (current.state === "pending" || current.state === "retry" || current.state === "active") &&
        event.previousState === current.state;
    case "retried":
      return (current.state === "failed" || current.state === "dead") &&
        event.previousState === current.state;
    case "dead":
      return (
        current.state === "active" ||
        current.state === "retry" ||
        current.state === "failed" ||
        current.state === "expired"
      ) && event.previousState === current.state;
    case "dismissed":
      return current.state === "dead" && event.previousState === "dead";
  }
}

function seedJobFromEvent<TPayload = unknown, TResult = unknown>(
  event: JobEvent<TPayload, TResult>,
): Job<TPayload, TResult> | undefined {
  if (event.payload === undefined) {
    return undefined;
  }

  return {
    id: event.jobId,
    service: event.service,
    type: event.jobType,
    state: event.state,
    payload: event.payload,
    createdAt: event.timestamp,
    updatedAt: event.timestamp,
    tries: event.tries,
    maxTries: event.maxTries ?? 1,
    ...(event.deadline !== undefined ? { deadline: event.deadline } : {}),
  };
}

File: rust/crates/trellis-jobs/src/projection.rs
use crate::types::{Job, JobEvent, JobEventType, JobState};

pub fn is_terminal(state: JobState) -> bool {
    matches!(
        state,
        JobState::Completed
            | JobState::Failed
            | JobState::Cancelled
            | JobState::Expired
            | JobState::Dead
            | JobState::Dismissed
    )
}

pub fn job_from_work_event(event: &JobEvent) -> Option<Job> {
    match event.event_type {
        JobEventType::Created | JobEventType::Retried => seed_job_from_event(event),
        _ => None,
    }
}

pub fn reduce_job_event(current: Option<&Job>, event: &JobEvent) -> Option<Job> {
    let current = match current {
        Some(value) => value,
        None => {
            if event.event_type != JobEventType::Created {
                return None;
            }
            return seed_job_from_event(event);
        }
    };

    if !is_legal_transition(current, event) {
        return Some(current.clone());
    }

    let mut next = Job {
        id: current.id.clone(),
        service: current.service.clone(),
        job_type: current.job_type.clone(),
        state: event.state,
        payload: current.payload.clone(),
        result: current.result.clone(),
        created_at: current.created_at.clone(),
        updated_at: event.timestamp.clone(),
        started_at: current.started_at.clone(),
        completed_at: current.completed_at.clone(),
        tries: event.tries,
        max_tries: event.max_tries.unwrap_or(current.max_tries),
        last_error: current.last_error.clone(),
        deadline: current.deadline.clone(),
        progress: current.progress.clone(),
        logs: current.logs.clone(),
    };

    match event.event_type {
        JobEventType::Started => {
            next.started_at = Some(event.timestamp.clone());
        }
        JobEventType::Progress => {
            next.progress = event.progress.clone();
        }
        JobEventType::Logged => {
            let mut logs = next.logs.take().unwrap_or_default();
            if let Some(new_logs) = &event.logs {
                logs.extend(new_logs.iter().cloned());
            }
            next.logs = Some(logs);
        }
        JobEventType::Completed => {
            next.result = event.result.clone();
            next.completed_at = Some(event.timestamp.clone());
        }
        JobEventType::Failed
        | JobEventType::Cancelled
        | JobEventType::Expired
        | JobEventType::Dead
        | JobEventType::Dismissed => {
            next.last_error = event.error.clone();
            next.completed_at = Some(event.timestamp.clone());
        }
        JobEventType::Retry => {
            next.last_error = event.error.clone();
        }
        JobEventType::Retried => {
            if let Some(payload) = &event.payload {
                next.payload = payload.clone();
            }
            if let Some(deadline) = &event.deadline {
                next.deadline = Some(deadline.clone());
            }
            next.result = None;
            next.completed_at = None;
            next.started_at = None;
            next.last_error = None;
            next.progress = None;
            next.logs = None;
            next.tries = 0;
        }
        JobEventType::Created => {}
    }

    Some(next)
}

fn is_legal_transition(current: &Job, event: &JobEvent) -> bool {
    match event.event_type {
        JobEventType::Created => false,
        JobEventType::Started => {
            matches!(current.state, JobState::Pending | JobState::Retry)
                && event.previous_state == Some(current.state)
        }
        JobEventType::Retry => {
            current.state == JobState::Active && event.previous_state == Some(JobState::Active)
        }
        JobEventType::Progress | JobEventType::Logged | JobEventType::Completed => {
            current.state == JobState::Active && event.previous_state == Some(JobState::Active)
        }
        JobEventType::Failed => {
            current.state == JobState::Active && event.previous_state == Some(JobState::Active)
        }
        JobEventType::Cancelled => {
            matches!(
                current.state,
                JobState::Pending | JobState::Retry | JobState::Active
            ) && event.previous_state == Some(current.state)
        }
        JobEventType::Expired => {
            matches!(
                current.state,
                JobState::Pending | JobState::Retry | JobState::Active
            ) && event.previous_state == Some(current.state)
        }
        JobEventType::Retried => {
            matches!(current.state, JobState::Failed | JobState::Dead)
                && event.previous_state == Some(current.state)
        }
        JobEventType::Dead => {
            matches!(
                current.state,
                JobState::Active | JobState::Retry | JobState::Failed | JobState::Expired
            ) && event.previous_state == Some(current.state)
        }
        JobEventType::Dismissed => {
            current.state == JobState::Dead && event.previous_state == Some(JobState::Dead)
        }
    }
}

fn seed_job_from_event(event: &JobEvent) -> Option<Job> {
    let payload = event.payload.clone()?;
    Some(Job {
        id: event.job_id.clone(),
        service: event.service.clone(),
        job_type: event.job_type.clone(),
        state: event.state,
        payload,
        result: None,
        created_at: event.timestamp.clone(),
        updated_at: event.timestamp.clone(),
        started_at: None,
        completed_at: None,
        tries: event.tries,
        max_tries: event.max_tries.unwrap_or(1),
        last_error: None,
        deadline: event.deadline.clone(),
        progress: None,
        logs: None,
    })
}
